// mcproto/lib/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Emitter = void 0;
class Emitter {
    constructor() {
        this.handlers = new Map();
        this.removeListener = this.off;
    }
    on(event, handler) {
        let handlers = this.handlers.get(event);
        if (handlers)
            handlers.add(handler);
        else
            this.handlers.set(event, new Set([handler]));
        return {
            dispose: () => this.removeListener(event, handler)
        };
    }
    once(event, handler) {
        const listener = this.on(event, ((...args) => {
            handler(...args);
            listener.dispose();
        }));
        return listener;
    }
    off(event, handler) {
        const handlers = this.handlers.get(event);
        if (!handlers)
            return false;
        return handlers.delete(handler);
    }
    emit(event, ...args) {
        const handlers = this.handlers.get(event);
        if (!handlers || handlers.size == 0)
            return false;
        [...handlers].forEach(cb => cb(...args));
        return true;
    }
}
exports.Emitter = Emitter;
